core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -924,49 +939,74 @@ private[deploy] class Master(
    * Return the UI if successful, else None
    */
   private[master] def rebuildSparkUI(app: ApplicationInfo): Option[SparkUI] = {
+    val futureUI = asyncRebuildSparkUI(app)
+
+    // Wait until the UI is rebuilt or cancelled, log status occasionally for apps with many events
+    var waitSec = 2
+    do {
+      try {
+        val ui = Await.result(futureUI, Duration(waitSec, TimeUnit.SECONDS))
+        return Some(ui)
+      } catch {
+        case te: TimeoutException =>
+          waitSec *= 2
+          logInfo(s"Application UI is rebuilding, will continue to wait $waitSec seconds")
+        case NonFatal(_) | _: InterruptedException =>
+          waitSec = 0
+      }
+    } while (waitSec > 0)
+    None
+  }
+
+  /** Rebuild a new SparkUI asynchronously to not block RPC event loop */
+  private[master] def asyncRebuildSparkUI(app: ApplicationInfo): Future[SparkUI] = {
     val appName = app.desc.name
     val notFoundBasePath = HistoryServer.UI_PATH_PREFIX + "/not-found"
-    try {
-      val eventLogDir = app.desc.eventLogDir
-        .getOrElse {
-          // Event logging is not enabled for this application
-          app.appUIUrlAtHistoryServer = Some(notFoundBasePath)
-          return None
-        }
+    val futureUI = Future {
+
+      // Fail the future immediately if event logging is not enabled for this application
+      val eventLogDir = app.desc.eventLogDir.getOrElse(throw new RebuildUINoEventLogException)
```

Comment:
we don't have to do this in the future since it's quick. Then we don't need to introduce the custom exception. I would actually just wrap the original code in a `Future[Option[SparkUI]`, and then `onSuccess` tell the RPC loop to attach it to the web UI.